functorch/experimental/_scan.py

```diff
@@ -0,0 +1,264 @@
+                             decompositions={},
+                             num_params_buffers=0,
+                             aot_id=0,
+                             keep_inference_input_mutations=False)
+
+
+def create_fw_bw_graph(f, flat_init, flat_xs):
+    # Note: We create "clean" environments for make_fx by suspending all dispatch keys
+    # between Autograd and Python key. Currently, we only suspend functionalization but more can be
+    # added when required. Will encounter two problems if we don't suspend functionalization:
+    #
+    # 1. make_fx fails to capture operations on input: the inputs are wrapped as _to_functional_tensor_wrapper,
+    # but they will be unwrapped before entering ProxyTorchDispatchMode as part of the dispatching.
+    # However, it's the outside wrapper that tracer creates proxies for. This casuses tracer fail to
+    # fetch the proxy for the inputs and fail to capture any operations on them.
+    #
+    # 2. make_fx fails to capture output: the outputs after ProxyTorchDispatchMode are further
+    # wrapped as FunctionalTensorWrapper in Functionalize key after return. However, the tracer
+    # only associates the inner tensor with proxy in ProxyTorchDispatchMode. Therefore,
+    # when creating the output node, it fails to associate the wrapped tensor with its proxy.
+    # Instead, it will create _tensor_constant as output.
+
+    with suspend_functionalization():
+        with disable_proxy_modes_tracing():
+            def from_fun(t):
+                if isinstance(t, torch.Tensor):
+                    return torch.empty_strided(t.size(), t.stride(), requires_grad=t.requires_grad)
+                return t
+
+            example_init = [from_fun(init_element) for init_element in flat_init]
+            example_xs = [from_fun(xs) for xs in _unstack_pytree(flat_xs)[0]]
+            example_flat_carry_out, example_flat_out = f(example_init, example_xs)
+            example_flat_carry_out, example_flat_out = pytree.tree_map(from_fun, example_flat_carry_out), pytree.tree_map(from_fun, example_flat_out)
+            if any(not isinstance(out, torch.Tensor) for out in example_flat_out if out is not None):
+                raise RuntimeError("Expect outputs of scan only contains tensors or None. "
+                                   f"Got types {[type(out) for out in example_flat_out]}.")
+            if any(not isinstance(out, torch.Tensor) for out in example_flat_carry_out if out is not None):
+                raise RuntimeError("Expect output carry of scan only contains tensors or None. "
+                                   f"Got types {[type(out) for out in example_flat_carry_out]}.")
+            example_grad_carry_out = [from_fun(out) for out in example_flat_carry_out]
+            example_grad_out = [from_fun(out) for out in example_flat_out]
+
+            fw_graph = make_fx(f)(example_init, example_xs)
+
+        def joint_f(example_grad_carry_out, example_xs, example_init, example_grad_out):
```

Comment:
We need to make this joint_f to take flat arguments so that In L97, make_fx will return a graph module that has no pytree logic in it. We may achieve this by recording the length of tuples.